README.md
# Test::Clustericious::Cluster [![Build Status](https://secure.travis-ci.org/clustericious/Test-Clustericious-Cluster.png)](http://travis-ci.org/clustericious/Test-Clustericious-Cluster)

(Deprecated) Test an imaginary beowulf cluster of Clustericious services

# SYNOPSIS

    use Test2::Bundle::More;
    use Test::Clustericious::Cluster;
    
    # suppose MyApp1 isa Clustericious::App and
    # MyApp2 is a Mojolicious app
    my $cluster = Test::Clustericious::Cluster->new;
    $cluster->create_cluster_ok('MyApp1', 'MyApp2');
    
    my @urls = @{ $cluster->urls };
    my $t = $cluster->t; # an instance of Test::Mojo
    
    $t->get_ok("$url[0]/arbitrary_path");  # tests against MyApp1
    $t->get_ok("$url[1]/another_path");    # tests against MyApp2
    
    done_testing;
    
    __DATA__
    
    @@ etc/MyApp1.conf
    ---
    # Clustericious configuration 
    url: <%= cluster->url %>
    url_for_my_app2: <%= cluster->urls->[1] %>

# DESCRIPTION

**NOTE**: This module has been deprecated, and may be removed on or after 31 December 2018.
Please see [https://github.com/clustericious/Clustericious/issues/46](https://github.com/clustericious/Clustericious/issues/46).

This module allows you to test an entire cluster of Clustericious services
(or just one or two).  The only prerequisites are [Mojolicious](https://metacpan.org/pod/Mojolicious), and
[Test2::Plugin::FauxHomeDir](https://metacpan.org/pod/Test2::Plugin::FauxHomeDir) so you can mix and match [Mojolicious](https://metacpan.org/pod/Mojolicious),
[Mojolicious::Lite](https://metacpan.org/pod/Mojolicious::Lite) and full [Clustericious](https://metacpan.org/pod/Clustericious) apps and test how they interact.

If you are testing against Clustericious applications, it is important to
either use this module as early as possible, or use [Test2::Plugin::FauxHomeDir](https://metacpan.org/pod/Test2::Plugin::FauxHomeDir)
as the very first module in your test, as testing Clustericious configurations
depend on the testing home directory being setup by [Test2::Plugin::FauxHomeDir](https://metacpan.org/pod/Test2::Plugin::FauxHomeDir).

In addition to passing [Clustericious](https://metacpan.org/pod/Clustericious) configurations into the
`create_cluster_ok` method as describe below, you can include configuration
in the data section of your test script.  The configuration files use 
[Clustericious::Config](https://metacpan.org/pod/Clustericious::Config), so you can use [Mojo::Template](https://metacpan.org/pod/Mojo::Template) directives to 
embed Perl code in the configuration.  You can access the [Test::Clustericious::Cluster](https://metacpan.org/pod/Test::Clustericious::Cluster)
instance from within the configuration using the `cluster` function, which
can be useful for getting the URL for the your and other service URLs.

    __DATA__
    
    @@ etc/Foo.conf
    ---
    url <%= cluster->url %>
    % # because YAML is (mostly) a super set of JSON you can
    % # convert perl structures into config items using json
    % # function:
    % # (json method requires Clustericious::Config 0.25)
    other_urls: <%= json [ @{ cluster->urls } ] %>

You can also put perl code in the data section of your test file, which
can be useful if there isn't a another good place to put it.  This
example embeds as [Mojolicious](https://metacpan.org/pod/Mojolicious) app "FooApp" and a [Clustericious::App](https://metacpan.org/pod/Clustericious::App)
"BarApp" into the test script itself:

    ...
    $cluster->create_cluster_ok('FooApp', 'BarApp');
    ...
    
    __DATA__
    
    @@ lib/FooApp.pm
    package FooApp;
    
    # FooApp is a Mojolicious app
    
    use Mojo::Base qw( Mojolicious );
    
    sub startup
    {
      shift->routes->get('/' => sub { shift->render(text => 'hello there from foo') });
    }
    
    1;
    
    @@ lib/BarApp.pm
    package BarApp;
    
    # BarApp is a Clustericious::App
    
    use strict;
    use warnings;
    use base qw( Clustericious::App );
    
    1;
    
    @@ lib/BarApp/Routes.pm
    package BarApp::Routes;
    
    use strict;
    use warnings;
    use Clustericious::RouteBuilder;
    
    get '/' => sub { shift->render(text => 'hello there from bar') };
    
    1;

These examples are full apps, but you could also use this
feature to implement mocks to test parts of your program
that use resources that aren't easily available during
unit testing, or may change from host to host.  Here is an
example that mocks parts of [Net::hostent](https://metacpan.org/pod/Net::hostent):

    use strict;
    use warnings;
    use Test::Clustericious::Cluster;
    use Test2::Bundle::More;
    
    eval q{ use Net::hostent };
    is $@, '';
    
    is gethost('bar')->name, 'foo.example.com', 'gethost(bar).name = foo.example.com';
    
    done_testing;
    
    __DATA__
    
    @@ lib/Net/hostent.pm
    package Net::hostent;
    
    use strict;
    use warnings;
    use base qw( Exporter );
    our @EXPORT = qw( gethost );
    
    sub gethost
    {
      my $input_name = shift;
      return unless $input_name =~ /^(foo|bar|baz|foo.example.com)$/;
      bless {}, 'Net::hostent';
    }
    
    sub name { 'foo.example.com' }
    sub aliases { qw( foo.example.com foo bar baz ) }
    
    1;

# CONSTRUCTOR

## new

    my $cluster = Test::Clustericious::Cluster->new( %args )

Arguments:

### t

The Test::Mojo object to use.
If not provided, then a new one will be created.

### lite\_path

List reference of paths to search for [Mojolicious::Lite](https://metacpan.org/pod/Mojolicious::Lite)
or PSGI apps.

# ATTRIBUTES

## t

    my $t = $cluster->t;

The instance of Test::Mojo used in testing.

## urls

    my @urls = @{ $cluster->urls };

The URLs for the various services.
Returned as an array ref.

## apps

    my @apps = @{ $cluster->apps };

The application objects for the various services.
Returned as an array ref.

## index

    my $index = $cluster->index;

The index of the current app (used from within a 
[Clustericious::Config](https://metacpan.org/pod/Clustericious::Config) configuration.

## url

    my $url = $cluster->url;

The url of the current app (used from within a
[Clustericious::Config](https://metacpan.org/pod/Clustericious::Config) configuration.

## auth\_url

    my $url = $cluster->auth_url;

The URL for the PlugAuth::Lite service, if one has been started.

# METHODS

## create\_cluster\_ok

    $cluster->create_cluster_ok( @services )

Adds the given services to the test cluster.
Each element in the services array may be either

- string

    The string is taken to be the [Mojolicious](https://metacpan.org/pod/Mojolicious) or [Clustericious](https://metacpan.org/pod/Clustericious)
    application class name.  No configuration is created or passed into
    the App.

    This can also be the name of a [Mojolicious::Lite](https://metacpan.org/pod/Mojolicious::Lite) application.
    The PATH environment variable will be used to search for the
    lite application.  The script for the lite app must be executable.
    You can specify additional directories to search using the
    `lite_path` argument to the constructor.

    This can also be a PSGI application.  In this case it needs to be
    in the `__DATA__` section of your test and it must have a name
    in the form `script/app.psgi`.  This also requires
    [Mojolicious::Plugin::MountPSGI](https://metacpan.org/pod/Mojolicious::Plugin::MountPSGI) already be installed so if you
    use this feature make sure you declare that as a prereq.

- list reference in the form: \[ string, hashref \]

    The string is taken to be the [Mojolicious](https://metacpan.org/pod/Mojolicious) application name.
    The hashref is the configuration passed into the constructor
    of the app.  This form should NOT be used for [Clustericious](https://metacpan.org/pod/Clustericious)
    apps (see the third form).

- list reference in the form: \[ string, string \]

    The first string is taken to be the [Clustericious](https://metacpan.org/pod/Clustericious) application
    name.  The second string is the configuration in either YAML
    or JSON format (may include [Mojo::Template](https://metacpan.org/pod/Mojo::Template) templating in it,
    see [Clustericious::Config](https://metacpan.org/pod/Clustericious::Config) for details).  This form requires
    that you have [Clustericous](https://metacpan.org/pod/Clustericous) installed, and of course should
    not be used for non-[Clustericious](https://metacpan.org/pod/Clustericious) [Mojolicious](https://metacpan.org/pod/Mojolicious) applications.

## create\_plugauth\_lite\_ok

    $cluster->create_plugauth_lite_ok( %args )

Add a [PlugAuth::Lite](https://metacpan.org/pod/PlugAuth::Lite) service to the test cluster.  The
`%args` are passed directly into the [PlugAuth::Lite](https://metacpan.org/pod/PlugAuth::Lite)
constructor.

You can retrieve the URL for the [PlugAuth::Lite](https://metacpan.org/pod/PlugAuth::Lite) service
using the `auth_url` attribute.

This feature requires [PlugAuth::Lite](https://metacpan.org/pod/PlugAuth::Lite) and [Clustericious](https://metacpan.org/pod/Clustericious) 
0.9925 or better, though neither are a prerequisite of this
module.  If you are using this method you need to either require
[PlugAuth::Lite](https://metacpan.org/pod/PlugAuth::Lite) and [Clustericious](https://metacpan.org/pod/Clustericious) 0.9925 or better, or skip 
your test in the event that the user has an earlier version. 
For example:

    use strict;
    use warnings;
    use Test::Clustericious::Cluster;
    use Test2::Bundle::More;
    BEGIN {
      skip_all 'test requires Clustericious 0.9925'
        unless eval q{ use Clustericious 1.00; 1 };
      skip_all 'test requires PlugAuth::Lite'
        unless eval q{ use PlugAuth::Lite 0.30; 1 };
    };

## stop\_ok

    $cluster->stop_ok( $index );
    $cluster->stop_ok( $index, $test_name);

Stop the given service.  The service is specified by 
an index, the first application when you created the
cluster is 0, the second is 1, and so on.

See [CAVEATS](https://metacpan.org/pod/Test::Clustericious::Cluster#CAVEATS)
below on interactions with IPv6 or TLS/SSL.

## start\_ok

    $cluster->start_ok( $index );
    $cluster->start_ok( $index, $test_name );

Start the given service.  The service is specified by 
an index, the first application when you created the
cluster is 0, the second is 1, and so on.

## is\_stopped

    $cluster->is_stopped( $index );
    $cluster->is_stopped( $index, $test_name );

Passes if the given service is stopped.

## isnt\_stopped

    $cluster->isnt_stopped( $index );
    $cluster->isnt_stopped( $index, $test_name );

Passes if the given service is not stopped.

## create\_ua

    my $ua = $cluster->create_ua;

Create a new instance of Mojo::UserAgent which can be used
to connect to nodes in the test cluster.

## extract\_data\_section

    $cluster->extract_data_section($regex);
    Test::Clustericious::Cluster->extract_data_section($regex);

Extract the files from the data section of the current package
that match the given regex.  `$regex` can also be a plain
string for an exact filename match.

## client

    my $client = $cluster->client($n);

Return a [Clustericious::Client](https://metacpan.org/pod/Clustericious::Client) object for use with the `$n`th
service in the cluster.  If there is a corresponding `YourService::Client`
class then it will be used.  Otherwise you will get a generic
[Clustericious::Client](https://metacpan.org/pod/Clustericious::Client) object with the correct URL configured.

This method only works with [Clustericious](https://metacpan.org/pod/Clustericious) services.

# CAVEATS

Some combination of Mojolicious, FreeBSD, IPv6 and TLS/SSL
seem to react badly to the use of 
[stop\_ok](https://metacpan.org/pod/Test::Clustericious::Cluster#stop_ok).  The work
around is to turn IPv6 and TLS/SSL off in the beginning
of any tests that uses stop\_ok your test like thus:

    use strict;
    use warnings;
    BEGIN { $ENV{MOJO_NO_IPV6} = 1; $ENV{MOJO_NO_TLS} = 1; }
    use Test::Clustericious::Cluster;

A proper fix would be desirable, see 

https://github.com/plicease/Test-Clustericious-Cluster/issues/3

If you want to help.

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.38      2018-03-12 19:06:20 -0700
  - Deprecated. See https://github.com/clustericious/Clustericious/issues/46

0.37      2017-07-22 23:42:49 -0400
  - Remove dependency on File::HomeDir

0.36_01   2017-07-14 11:10:19 -0400
  - Migrate to using Test2::V0

0.35      2017-04-27 08:37:00 -0400
  - Bump Clustericious tests to requrie 1.24 for Mojolicious 7.31 compat

0.34      2017-03-14 10:59:52 -0400
  - Migrate to using Test2 + Test2::Suite

0.33      2016-05-29 10:27:38 +0200
  - You can now explicitly add .psgi extension to PSGI apps.

0.32      2016-05-12 19:24:50 -0400
  - Also search for .psgi files in the "lite_path"
  - Always create the "cluster" config helper if
    Clustericious is installed.
  - Fixed bug where constructor was ignoring Test::Mojo passed to it

0.31      2016-04-22 13:36:21 -0400
  - Mojolicious 6.58 compatability

0.30      2016-03-14 12:21:55 -0400
  - Clustericious 1.17 compatability

0.29      2016-01-03 10:35:00 -0500
  - Can now test a ::Client without creating a .conf file
  - Can now get a client object for each app

0.28      2015-12-04 11:34:38 -0500
  - Fix skip logic in test t/clustericious_config_with_colon.t

0.27      2015-12-04 09:41:10 -0500
  - Fix bug where an app name with colons was looking for
    config in with the wrong name.  Note the config filename
    for app Foo::Bar is actually etc/Foo-Bar.conf.  This is
    because : is not legal in filenames on some platforms.

0.26      2015-12-04 08:03:53 -0500
  - Add support for testing PSGI apps
    this requires Mojolicious::Plugin::MountPSGI
  - extract_data_section can take a plain string for
    an exact match of one file.
  - extract_data_section can be called as a class method
    (this was already the case, but it is documented/supported now)
  - Auto extract lib/* from data section during create_cluster_ok
    (we may auto extract lib/... and script/... files in the future
     as necessary)
  - Default to the last service for $cluster->t testing
    (ie $cluster->t->get_ok('/foo') should work instead of needing
        $cluster->t->get_ok("$url/foo") for single service clusters)
    This Feature should be considered experimental

0.25      2015-12-01 06:56:07 -0500
  - Fixed bug where create_cluster_ok not always catching failure on
    start of Clustericious App.

0.24      2015-11-30 08:27:20 -0500
  - Added extract_data_section method

0.23      2015-11-02 13:00:57 -0500
  - Fixed test skip logic

0.22      2015-11-02 08:40:32 -0500
  - Hook into Clustericious::Client if app is a Clustericious service
    on Clustericious 1.01 or better

0.21      2015-07-07 11:27:22 -0400
  - Track stops / starts
    This fixes a bug when you use create_ua after a service has been stopped
  - Fix long standing incompatability wiht recent Mojolicious and PlugAuth::Lite
    tests
  - added tests is_stopped and isnt_stopped

0.20      2015-02-28 10:54:59 -0500
  - Mojo 6.x compat

0.19      2015-01-22 11:17:21 -0500
  - fixed borked tests if recent Clustericious is already installed.

0.18      2014-12-30 00:02:32 -0500
  - Clustericious test cases run only with 0.9941 or better
    this is the new combined Clustericious coming out tomorrow.

0.17      2014-12-19 12:45:10 -0500
  - merge former optional dependency PlugAuth::Lite into this distribution
  - Do not use Mojo::JSON deprecated OO interface

0.16      2014-05-30 09:39:01 -0400
  - Mojolicious 5.x compatability

0.15      2014-05-02 19:26:51 -0400
  - promote to production

0.14_01   2014-04-18 08:34:05 -0400
  - Experimental: better detection in t/stop.t

0.14      2014-04-17 03:26:32 -0400
  - remove experimental dep on Carp::Always / Carp::Always::Dump
  - remove some experimental diagnostics
  - promote the rest to production

0.12_07   2014-04-16 18:22:31 -0400
  - Experimental: turn off IPv6 and TLS support in t/stop.t

0.12_06   2014-04-16 14:11:20 -0400
  - Experimental: additional diagnostics

0.12_05   2014-04-10 13:25:54 -0400
  - Experimental: fixed additional diagnostics
  - Use Module::Build instead of MakeMaker

0.12_04   2014-04-10 05:44:00 -0400
  - Experimental: additional diagnostics

0.12_03   2014-04-09 06:38:41 -0400
  - Experimental: explicitly stop server before undefing in stop_ok

0.12_02   2014-04-08 19:55:32 -0400
  - Experimental: temp use Carp::Always for additional diagnostic

0.12_01   2014-04-08 12:09:19 -0400
  - Experimental: catch error in stop and report as stop_ok failure

0.12      2013-10-27 04:15:58 -0400
  - Mojolicious 4.50 compatability

0.11      2013-10-17 22:00:17 -0400
  - rm undocumented and never used private _builder attribute

0.10      2013-09-03 18:14:15 -0400
  - Clustericious specific tests require Clustericious 0.9930
    (they are skipped if not installed)

0.09      2013-07-22 06:30:30 America/New_York
  - make it work with a mojo lite app that returns app->start

0.08      2013-07-19 22:50:51 America/New_York
  - promote to production

0.07_01   2013-07-19 22:46:59 America/New_York
  - work around the limitations of case insensative filesystems
    by trying YourApp before lite app yourapp

0.07      2013-07-19 16:17:12 America/New_York
  - promote to production

0.06_01   2013-07-18 13:45:53 America/New_York
  - MSWin32 compatability fixes

0.06      2013-07-18 09:09:06 America/New_York
  - promote to production

0.05_01   2013-07-17 16:36:31 America/New_York
  - documentation
  - inline perl code gets put in front of @INC

0.05      2013-07-14 22:33:34 America/New_York
  - support for non-inlined lite apps
  - deprecate passing Test::Mojo instance into new

0.04      2013-07-07 00:39:12 America/New_York
  - patch %INC to make Mojo::Home happy
  - create_plugauth_lite_ok requires Clustericious 0.9925
    this was always true, but this version:
  - document dep on Clustericious 0.9925
  - croak if you try to use create_plugauth_lite_ok with
    an earlier version

0.03      2013-07-05 22:27:24 America/New_York

lib/Test/PlugAuth.pm
package Test::PlugAuth;

use strict;
use warnings;
use 5.010001;
use PlugAuth::Lite;
use Mojo::UserAgent;

# ABSTRACT: minimum PlugAuth server to test Clustericious apps against
# VERSION

=head1 SYNOPSIS

assuming you have a Clustericious app MyApp with authentication/authorization
directives that you need to test:

 use Test::Clustericious::Config;
 use Test::Clustericious;
 use Test::PlugAuth;
 
 my $auth = Test::PlugAuth->new(auth => {
   my($user,$pass) = @_;
   return $user eq 'gooduser' && $pass eq 'goodpass';
 });
 
 create_config_ok 'MyApp', { plug_auth => { url => $auth->url } };
 
 $t = Test::Clustericious->new('MyApp');
 $auth->apply_to_client_app($t->app);
 
 my $port = $t->ua->app_url->port;
 
 $t->get_ok("http://baduser:badpass\@localhost:$port/private")
   ->status_is(401);
 $t->get_ok("http://gooduser:goodpass\@localhost:$port/private")
   ->status_is(200);

=head1 DESCRIPTION

Provides a way to test a Clustericious application with a fake PlugAuth server
with reduced boilerplate

=head1 CONSTRUCTOR

=head2 Test::PlugAuth->new( $config )

Creates a new instance of Test::PlugAuth.  The $config is passed
directly into L<PlugAuth::Lite>.  See L<Mojolicious::Plugin::PlugAuthLite>
for details.

=cut

sub new
{
  my $class = shift;
  my $config = ref $_[0] ? $_[0] : {@_};
  my $self = bless {}, $class;
  
  $self->{app} = PlugAuth::Lite->new($config);
  $self->{ua}  = Mojo::UserAgent->new;
  eval { $self->ua->server->app($self->app) } // $self->ua->app($self->app);
  
  $self->{url} = eval { $self->ua->server->url->to_string } // $self->ua->app_url->to_string;
  $self->{url} =~ s{/$}{};
  
  return $self;
}

=head1 ATTRIBUTES

=head2 ua

The L<Mojo::UserAgent> used to connect to the PlugAuth (lite) server.

=cut

sub ua  { shift->{ua}  }

=head2 app

The L<PlugAuth::Lite> instance of the PlugAuth server.

=cut

sub app { shift->{app} }

=head2 url

The (fake) url used to connect to the PlugAuth server with.  You MUST
connect to through the L<Mojo::UserAgent> above.

=cut

sub url { shift->{url} }

=head1 METHODS

=head2 $test_auth->apply_to_client_app( $client_app )

Given a Clustericious application C<$client_app>, this method will 
rewire our L<Mojo::UserAgent> for authentication requests to PlugAuth.

=cut

sub apply_to_client_app 
{
  my($self, $client_app) = @_;
  $client_app->helper(auth_ua => sub { $self->ua });
  return;
}

1;


lib/PlugAuth/Lite.pm
package PlugAuth::Lite;

use strict;
use warnings;
use 5.010001;
use Mojo::Base qw( Mojolicious );

# ABSTRACT: Pluggable (lite) authentication and authorization server.
# VERSION

=head1 SYNOPSIS

command line:

 % plugauthlite

Mojolicious Plugin:

 use Mojolicious::Lite;
 
 plugin 'plug_auth_lite', 
   auth => sub {
     my($user, $pass) = @_;
     if($user eq 'optimus' && $pass eq 'matrix')
     { return 1; }
     else
     { return 0; }
   },
   authz => sub {
     my($user, $action, $resource) = @_;
     if($user eq 'optimus && $action eq 'open' && $resource =~ m{^/matrix})
     { return 1 }
     else
     { return 0 }
   };

Mojolicious App:

 use PlugAuth::Lite;
 my $app = PlugAuth::Lite->new({
   auth => sub {
     my($user, $pass) = @_;
     if($user eq 'optimus' && $pass eq 'matrix')
     { return 1; }
     else
     { return 0; }
   },
   authz => sub {
     my($user, $action, $resource) = @_;
     if($user eq 'optimus && $action eq 'open' && $resource =~ m{^/matrix})
     { return 1 }
     else
     { return 0 }
   },
 });

=head1 DESCRIPTION

This distribution provides 

=over 4

=item L<plugauthlite>

L<PlugAuth> compatible server in the form of a L<Mojolicious::Lite> application.

=item L<Mojolicious::Plugin::PlugAuthLite>

L<Mojolicious> plugin that adds L<PlugAuth> compatible routes to a new or 
existing L<Mojolicious> application.

=item L<PlugAuth::Lite>

L<Mojolicious> application with L<PlugAuth> compatible routes that can be spawned
from within a perl application.

=back

In the future it will also contain a testing interface for testing authentication
and authorization rules in L<Clustericious> applications.

It has fewer prerequisites that the full fledged L<PlugAuth> server (simply
L<Mojolicious> and perl itself) but also fewer features (it notably lacks
the management interface).

=head1 ATTRIBUTES

=head2 auth

Subroutine reference to call to check authentication.  Passes in C<($user, $pass)> should
return true for authenticated, false otherwise.

If not provided, all authentications fail.

=head2 authz

Subroutine reference to call to check authorization.  Passes in C<($user, $action, $resource)>
and should return true for authorized, false otherwise.

If not provided, all authorizations succeed.

=head2 host

Subroutine reference to call to check host information.

=cut

has 'auth';
has 'authz';
has 'host';

sub startup
{
  my($self, $config) = @_;

  $self->plugin('plug_auth_lite',
    auth  => $self->auth  // sub { 0 },
    authz => $self->authz // sub { 1 },
    host  => $self->host  // sub { 0 },
  );
}

1;

=head1 SEE ALSO

L<plugauthlite>,
L<Mojolicious::Plugin::PlugAuthLite>,
L<Test::PlugAuth>,
L<Clustericious>

=cut



lib/Mojolicious/Plugin/PlugAuthLite.pm
package Mojolicious::Plugin::PlugAuthLite;

use Mojo::Base qw( Mojolicious::Plugin );
use Mojo::ByteStream qw( b );
use 5.010001;

# ABSTRACT: Add a minimal PlugAuth server to your Mojolicious application.
# VERSION

=head1 SYNOPSIS

 use Mojolicious::Lite
 
 plugin 'plug_auth_lite', 
   auth => sub {
     my($user, $pass) = @_;
     if($user eq 'optimus' && $pass eq 'matrix')
     { return 1; }
     else
     { return 0; }
   },
   authz => sub {
     my($user, $action, $resource) = @_;
     if($user eq 'optimus && $action eq 'open' && $resource =~ m{^/matrix})
     { return 1 }
     else
     { return 0 }
   };

=head1 DESCRIPTION

This plugin provides a very minimal but customizable L<PlugAuth> server which can
be included with your L<Mojolicious> application for L<Clustericious> applications
to authenticate against.  If you do not need specialized plugins for LDAP or DBI,
and if you do not need the user/group/resource management provided by a the full
featured L<PlugAuth> server then this plugin may be for you.

The script L<plugauthlite> included with this distribution provides PlugAuth
style authentication (but not authorization) using a simple Apache style password
file.

=head1 CONFIGURATION

=head2 auth

Subroutine which checks the authentication of a user.  It is passed two arguments,
the username and the password.  If they are authentic this call back should return
1.  Otherwise it should return 0.

=head2 authz

Subroutine which checks the authorization of a user.  It is passwd three arguments,
the username, action (usually a verb) and resource (usually the path part of a URL).
If the user is authorized for the action on that resource the call back should return
1.  Otherwise it should return 0.

=head2 url

The prefix to prepend to the standard PlugAuth API routes.  Usually the authentication
route is /auth and the authorization route is /authz, but if the PlugAuth.conf client
configuration is set to http://example.com/foo the client expects the authentication
route to be /foo/auth and the authorization route to be /foo/authz.  In this case you
would set this configuration item to '/foo'.

=head2 realm

The realm to use for HTTP Basic authentication.  The default is PlugAuthLite.

=head1 ROUTES

=head2 GET /auth

=over 4

=item * if username and password provided using BASIC authentication and are correct

Return 200 ok

=item * if username and password provided using BASIC authentication but are not correct

Return 403 not ok

=item * if username and password are not provided using BASIC authentication

Return 401 please authenticate

=back

=head2 GET /authz/user/#user/#action/(*resource)

=over 4

=item * if the given user (#user) is permitted to perform the given action (#action) on the given resource (*resource)

Return 200 ok

=item * otherwise

return 403 not ok

=back

=head1 METHODS

=head2 register

This method adds the routes to your application required to implement the PlugAuth
API.

=head1 LIMITATIONS

This implementation of the PlugAuth protocol does not support these features provided
by the full fledged L<PlugAuth> server:

=over 4

=item *

Groups

=item *

Management API for creating/removing/modifying users/groups/resources

=item *

Standard Clustericious routes like "/version" and "/status"

=item *

Clustericious configuration file (~/etc/PlugAuth.conf)

=item *

Support for L<PlugAuth> plugins (L<PlugAuth::Plugin>).

=item *

Probably many others.

=back

=head1 SEE ALSO

L<plugauthlite>,
L<PlugAuth::Lite>,
L<PlugAuth>

=cut

sub register
{
  my($self, $app, $conf) = @_;
  
  my $cb_auth  = $conf->{auth}  // sub { 0 };
  my $cb_authz = $conf->{authz} // sub { 1 };
  my $cb_host  = $conf->{host}  // sub { 0 };
  my $realm = $conf->{realm} // 'PlugAuthLite';
  my $base_url = $conf->{url} // $conf->{uri} // '';

  $app->routes->get("$base_url/auth" => sub {
    my $self = shift;
    eval {
      my $auth_header = $self->req->headers->authorization;
      unless($auth_header)
      {
        $self->res->headers->www_authenticate("Basic \"$realm\"");
        $self->render(text => 'please authenticate', status => 401);
        return;
      }
      my ($method,$str) = split / /,$auth_header;
      my ($user,$pw) = split /:/, b($str)->b64_decode;
      if($cb_auth->($user, $pw))
      {
        $self->render(text => 'ok', status => 200);
      }
      else
      {
        $self->render(text => 'not ok', status => 403);
      }
    };
    $self->render(text => 'not ok', status => 503) if $@;
  })->name('plugauth_auth');
  
  $app->routes->get("$base_url/authz/user/#user/#action/(*resource)" => { resource => '/' } => sub {
    my $self = shift;
    eval {
      my($user, $resource, $action) = map { $self->stash($_) } qw( user resource action );
      $resource =~ s{^/?}{/};
      if($cb_authz->($user, $action, $resource))
      {
        $self->render(text => 'ok', status => 200);
      }
      else
      {
        $self->render(text => 'not ok', status => 403);
      }
    };
    $self->render(text => 'not ok', status => 503) if $@;
  })->name('plugauth_authz');
  
  $app->routes->get("$base_url/host/#host/:tag" => sub {
    my $self = shift;
    eval {
      my ($host,$tag) = map $self->stash($_), qw/host tag/;
      if ($cb_host->($host,$tag)) {
        return $self->render(text => 'ok', status => 200);
      }
      return $self->render(text => 'not ok', status => 403);
    };
    $self->render(text => 'not ok', status => 503) if $@;
  })->name('plugauth_host');
  
  return;
}

1;
